apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableReadRowsRequestManager.cs

```diff
@@ -0,0 +1,134 @@
+
+using System.Diagnostics;
+using Google.Protobuf;
+using System.Linq;
+using System;
+
+namespace Google.Cloud.Bigtable.V2
+{
+    /// <summary>
+    /// Keeps track of Rows returned from a readRows RPC for information relevant to resuming the RPC
+    /// after temporary problems.
+    /// </summary>
+    internal class BigtableReadRowsRequestManager
+    {
+        // Member variables from the constructor.
+        private readonly ReadRowsRequest _originalRequest;
+
+        // The number of rows read so far.
+        internal long RowCount { get; private set; }
+
+        /// <summary>
+        /// Constructor for <see cref="BigtableReadRowsRequestManager"/>.
+        /// </summary>
+        /// <param name="originalRequest">
+        /// Original ReadRowsRequest containing all of the parameters of the API call.
+        /// </param>
+        internal BigtableReadRowsRequestManager(ReadRowsRequest originalRequest) => _originalRequest = originalRequest;
+
+        /// <summary>
+        /// Update last found Rowkey
+        /// </summary>
+        internal BigtableByteString LastFoundKey { get; set; }
+
+        /// <summary>
+        /// Update amount of rows read so far
+        /// </summary>
+        internal void IncrementRowCount(int count)
+        {
+            RowCount += count;
+        }
+
+        /// <summary>
+        /// Builds and returns updated subrequest that excludes all rowKeys that have already been found.
+        /// </summary>
+        internal ReadRowsRequest BuildUpdatedRequest()
+        {
+            ReadRowsRequest newReadRowsRequest = new ReadRowsRequest
+            {
+                TableNameAsTableName = _originalRequest.TableNameAsTableName,
+            };
+            
+            // Transfer RowFilter.
+            newReadRowsRequest.Filter = _originalRequest.Filter;
+            
+            // If the row limit is set, update it.
+            if (_originalRequest.RowsLimit != 0)
+            {
+                long rowsRemaining = _originalRequest.RowsLimit - RowCount; 
+                Debug.Assert(rowsRemaining > 0, "The remaining number of rows must be greater than 0.");
+                newReadRowsRequest.RowsLimit = Math.Max(1, rowsRemaining);
+            }
+
+            // Remove received rows form RowSet.
+            newReadRowsRequest.Rows = FilterRows();
+
+            return newReadRowsRequest;
+        }
+
+        /// <summary>
+        /// Removes receied rows from the RowSet.
+        /// </summary>
+        /// <returns>
+        /// New RowSet with not received rows.</returns>
+        private RowSet FilterRows()
+        {
+            RowSet originalRows = _originalRequest.Rows;
+
+            if (LastFoundKey.Value.IsEmpty)
+            {
+                return originalRows;
+            }
+
+            RowSet newRowSet = new RowSet();
+
+            newRowSet.RowKeys.AddRange(originalRows.RowKeys.Where(key => !StartKeyIsAlreadyRead(key)));
```

Comment:
Then I'll rename `RowSet FilterRows()` to `RepeatedField<RowRange> FilterRowRanges` and `RowSet newRowSet = new RowSet` to `RepeatedField<RowRange> newRowRanges = _originalRequest.Rows.RowRanges` accordingly.

This way object creation expression will have 
```
TableNameAsTableName = _originalRequest.TableNameAsTableName,
// Transfer RowFilter.
Filter = _originalRequest.Filter,
// Remove received rowKeys form Rows.
Rows =
{
    RowKeys = {_originalRequest.Rows.RowKeys.Where(key => !StartKeyIsAlreadyRead(key))},
    RowRanges = {FilterRowRanges()}
}
```

Reply:
I'm not quite sure I follow, but I'll just look at the updated code. Also, you should be able to use `IEnumerable<...>` rather than `RepeatedField<...>` and use any collection or iterator block.